apps/framework-cli/src/framework/core/state_storage.rs

```diff
@@ -0,0 +1,390 @@
+pub struct MigrationLock {
+    pub machine_id: String,
+    pub started_at: DateTime<Utc>,
+    pub expires_at: DateTime<Utc>,
+}
+
+#[async_trait]
+pub trait StateStorage: Send + Sync {
+    /// Store the infrastructure map
+    async fn store_infrastructure_map(&self, infra_map: &InfrastructureMap) -> Result<()>;
+
+    /// Load the infrastructure map
+    async fn load_infrastructure_map(&self) -> Result<Option<InfrastructureMap>>;
+}
+
+/// Redis-based state storage
+pub struct RedisStateStorage {
+    client: Arc<RedisClient>,
+}
+
+impl RedisStateStorage {
+    pub fn new(client: Arc<RedisClient>) -> Self {
+        Self { client }
+    }
+}
+
+#[async_trait]
+impl StateStorage for RedisStateStorage {
+    async fn store_infrastructure_map(&self, infra_map: &InfrastructureMap) -> Result<()> {
+        infra_map.store_in_redis(&self.client).await
+    }
+
+    async fn load_infrastructure_map(&self) -> Result<Option<InfrastructureMap>> {
+        InfrastructureMap::load_from_last_redis_prefix(&self.client).await
+    }
+}
+
+/// ClickHouse-based state storage (for serverless/CLI-only deployments)
+pub struct ClickHouseStateStorage {
+    client: ConfiguredDBClient,
+    db_name: String,
+}
+
+impl ClickHouseStateStorage {
+    const STATE_TABLE: &'static str = "_MOOSE_STATE";
+    const LOCK_KEY: &'static str = "migration_lock";
+    const LOCK_TIMEOUT_SECS: i64 = 300; // 5 minutes
+
+    pub fn new(client: ConfiguredDBClient, db_name: String) -> Self {
+        Self { client, db_name }
+    }
+
+    /// Ensure the state table exists using KeeperMap for strong consistency
+    async fn ensure_state_table(&self) -> Result<()> {
+        // Use KeeperMap for:
+        // 1. Atomic lock operations (prevents concurrent migrations)
+        // 2. Synchronous writes (no async_insert race conditions)
+        // 3. Immediate read-after-write consistency
+        // 4. Already configured in dev mode. Available in Clickhouse Cloud
+        let create_table_sql = format!(
+            r#"
+            CREATE TABLE IF NOT EXISTS `{}`.`{}`
+            (
+                key String,
+                value String,
+                created_at DateTime DEFAULT now()
+            )
+            ENGINE = KeeperMap('/{}/{}')
+            PRIMARY KEY key
+            "#,
+            self.db_name,
+            Self::STATE_TABLE,
+            self.db_name,
+            Self::STATE_TABLE
+        );
+
+        debug!("Creating KeeperMap state table: {}", create_table_sql);
+
+        self.client
+            .client
+            .query(&create_table_sql)
+            .execute()
+            .await
+            .context("Failed to create state table")?;
+
+        Ok(())
+    }
+
+    /// Try to acquire migration lock
+    /// Must be manually released with release_migration_lock()
+    /// Lock automatically expires after 5 minutes as a safety fallback
+    pub async fn acquire_migration_lock(&self) -> Result<()> {
```

Comment:
Ohhhh - Would be so good to have that on the redis side too. 

Not a blocker but that would enable to make the Redis Client a lot more stateless. No more subscribing to redis and keep alive.